### ISSUE TITLE
Radial Gradient Layer

### ISSUE THREAD

**Issue by Anywhere-Music-Player** [cocos2d/cocos2d-x#17480]
Hello,

I'm looking for a RadialGradientLayer for cocos2d-x. 
Discussion started here http://discuss.cocos2d-x.org/t/radial-gradient/26602/6

I don't know OpenGL and shaders, so I can't implement it in any way. Really looking for solution from cocos2d-x developers. 

Some helpful links from thread: 
https://github.com/saeedafshari/ShaderLayer 
https://github.com/SFML/SFML/wiki/Source:-Radial-Gradient-Shader

Can this feature be added to cocos2d-x? I believe this should be easy to implement for cocos2d-x developers..

**Comment by minggo** [comment]
The post you provide has the shader, you can use them in your project. If adding a feature in cocos2d-x, then we may need to make it as common as possible, such as should change from one color to another one? Should support multiple colors? If you only need the feature as you posted above, i think i can give you an example work in cocos2d-x.

**Comment by volodymyr-vestasoft** [comment]
I'm just looking for almost same as LayerGradient. So, with typical options like:
 - startColor
 - endColor
 - startOpacity
 - endOpacity

But for radial, LayerRadialGradient I believe should be added a few more options:
 - radius
 - expand
 - center(optional), should be like always (0.5,0.5) in a node as I think

**Comment by Anywhere-Music-Player** [comment]
I'm just looking for almost same as LayerGradient. So, with typical options like:
 - startColor
 - endColor
 - startOpacity
 - endOpacity

But for radial, LayerRadialGradient I believe should be added a few more options:
 - radius
 - expand
 - center(optional),  I think should be by default (0.5,0.5), in a node center.

**Comment by bsy6766** [comment]
Shader can be an easier way and actually might be a better way to draw smooth circle, plus messing around with colors, compared to polygon circle.
I followed your shader example with LayerColor and the result was like [this](http://imgur.com/a/fVqnh).

**Comment by minggo** [comment]
@KAMIKAZEUA what does `expand ` used for?

**Comment by Anywhere-Music-Player** [comment]
@minggo full visual explanation you can find here https://github.com/SFML/SFML/wiki/Source:-Radial-Gradient-Shader

**Comment by minggo** [comment]
I think we don't need expand since we have `startColor` and `endColor`.

**Comment by minggo** [comment]
Oh, it seems expand is used to make inner color. But what's the inner color, it is the startColor in the case?

**Comment by Anywhere-Music-Player** [comment]
Yes, inner it's startColor. And gives a bit more flexibility to configure gradient.

**Comment by minggo** [comment]
Sounds reasonable. Thanks @KAMIKAZEUA .

**Comment by Anywhere-Music-Player** [comment]
Looking forward for implementation :)

**Comment by minggo** [comment]
@KAMIKAZEUA i sent https://github.com/cocos2d/cocos2d-x/pull/17826 for you to do testing. You can merge the PR into your codes and have a try. I modified `HelloWorldScene.cpp` of `cpp-empty-test` to do testing, the codes looks like:
```c++
Vec2 center(Vec2(visibleSize / 2) + origin);
    auto layerRadialGradient = LayerRadialGradient::create(Color4B::RED, Color4B::BLUE, 50.0f, center, 0.5f);
    this->addChild(layerRadialGradient);
```

And the effect looks like following on mac:
<img width="242" alt="screen shot 2017-05-17 at 10 22 30 am" src="https://cloud.githubusercontent.com/assets/488203/26135837/c57668ac-3aea-11e7-8baa-30e75501ae38.png">

I think can not pass `startOpacity` and `endOpacity` when creating, because they are part of `startColor` and `endColor`, but can modify them separately after creating.

I will complete the codes after your feedback. 
Thanks.

**Comment by minggo** [comment]
@KAMIKAZEUA any feedback?
@bsy6766 how did you implement it with LayerColor? Will you please also give some feedback too?

Thanks.

**Comment by Anywhere-Music-Player** [comment]
@minggo my time zone a bit diff from yours :) so, I just started to work a couple of hours ago on my main work. In 8h since now, I would love to test this asap when I get home. This is really cool thing what I need for my game, can't wait1! 

p.s. its called "...works on mac and iOS" but on android it will work too? This thing needed for my cross-platform game :)

**Comment by minggo** [comment]
> p.s. its called "...works on mac and iOS" but on android it will work too? This thing needed for my cross-platform game :)

No. When i sent the PR, i just tested on iOS and Mac. Now i tested on Android and windows too, so it should work on all supported platforms.

**Comment by G0561** [comment]
cocos2dx 3.13  下，这么段代码，本身点移动是从右上角往左下角，但是现在确实往右上角移动，不知道为什么会这样，求解答


DrawNode* dot = DrawNode::create(5.0);
dot->drawDot(Vec2(visibleSize.width*0.8, visibleSize.height*0.8), 50, Color4F::GREEN);
this->addChild(dot);
ActionInterval * move1 = MoveTo::create(2.0, Vec2(origin.x + visibleSize.width*0.5, origin.y + visibleSize.height*0.5));
dot->runAction(move1);
这段代码， drawnode 的图形，移动会有问题，上面是从右上角往左下角移动，但方向确实相反的








At 2017-05-17 17:37:35, "minggo" <notifications@github.com> wrote:


p.s. its called "...works on mac and iOS" but on android it will work too? This thing needed for my cross-platform game :)

No. When i sent the PR, i just tested on iOS and Mac. Now i tested on Android and windows too, so it should work on all supported platforms.

**Comment by Anywhere-Music-Player** [comment]
@minggo I just tested it on iOS device and Mac. All looks good for me 👍 . But I wanted to change startOpacity. 

I believe it should work like that: 

`LayerRadialGradient::create(Color4B(145, 106, 209, 150),Color4B(0, 0, 0, 0), 250.0f, center, 0.0f);`

Also, startOpacity is private.

**Comment by minggo** [comment]
@KAMIKAZEUA yep, as i said, i will complete them when the effect is good.

> I believe it should work like that:
>
> LayerRadialGradient::create(Color4B(145, 106, 209, 150),Color4B(0, 0, 0, 0), 250.0f, center, 0.0f);

What did you mean?

@G0561 i think you add comment in wrong issue.

**Comment by minggo** [comment]
@KAMIKAZEUA i complete the work and move the test case to `cpp-tests/LayerTest`.

**Comment by Anywhere-Music-Player** [comment]
I mean that passing color Color4B(145, 106, 209, **150**) with opacity value is this startOpacity value. 

My target image actually this: 

![image](https://cloud.githubusercontent.com/assets/3391456/26189435/5ad4b62a-3bad-11e7-9e8d-0f36e37e5d7d.png)

I just checked cpp-tests/LayerTest. 
Q: is there an option to chance whole opacity for layer? Like `_layer->setOpacity(150);`

**Comment by minggo** [comment]
@KAMIKAZEUA i think you can take a look of `LayerRadialGradient::setBlendFunc()` . I don't know what's your requirements, or is the effect is wrong? Could you please describe it in more detail?
Thanks.

**Comment by Anywhere-Music-Player** [comment]
Yes, sure. I'm converting my game written in Objc and here is method I've used to create my radial gradient for iOS: 

```
- (UIImage *)radialGradientImageWithRadius:(CGFloat)radius StartColor:(UIColor*)startColor EndColor:(UIColor*)endColor
{
    // Initialize
    UIGraphicsBeginImageContextWithOptions(CGSizeMake(radius, radius), NO, 1.0);
    
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    // bottom glow gradient
    CGColorSpaceRef colourspace = CGColorSpaceCreateDeviceRGB();
    
    // build color components
    CGFloat red1 = 0.f, green1 = 0.f, blue1 = 0.f, alpha1 = 0.f;
    [(startColor == nil ? [UIColor clearColor] : startColor) getRed:&red1 green:&green1 blue:&blue1 alpha:&alpha1];
    CGFloat red2 = 0.f, green2 = 0.f, blue2 = 0.f, alpha2 = 0.f;
    [(endColor == nil ? [UIColor clearColor] : endColor) getRed:&red2 green:&green2 blue:&blue2 alpha:&alpha2];
    
    CGFloat cComponents[] = { red1, green1, blue1, alpha1, red2, green2, blue2, alpha2};
    CGFloat cGlocations[] = { 0.f, 1.0f };
    
    CGGradientRef gradient = CGGradientCreateWithColorComponents(colourspace, cComponents, cGlocations, 2);
    CGPoint centerPoint = CGPointMake(radius * 0.5, radius * 0.5);
    
    CGContextDrawRadialGradient(context, gradient, centerPoint, 0.f, centerPoint, radius * 0.5 , 0.f);
    
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    
    CGGradientRelease(gradient);
    CGColorSpaceRelease(colourspace);
    UIGraphicsEndImageContext();
    
    return image;
}
```

```
UIImage *img = [self radialGradientImageWithRadius: 873.0 
                                        StartColor: [UIColor colorWithRed:145.0/255.0 green:106.0/255.0 blue:209.0/255.0 alpha:0.55]
                                          EndColor: nil];
```

Output was as I posted image before. Current effect looks similar, but not really exactly the same.

I rendered it to check(I've set opacity to 180, but with old options it's should be 140, but it's too light with current implementation): 
```
    _layer = LayerRadialGradient::create(Color4B(145, 106, 209, 180),Color4B(0, 0, 0, 0), 350.0f, center, 0.0f);
    
    RenderTexture *render = RenderTexture::create(size.width, size.height);
    render->begin();
    _layer->visit();
    render->end();
    render->saveToFile("123.png", Image::Format::PNG);
```
![123](https://cloud.githubusercontent.com/assets/3391456/26190471/560f910a-3bb2-11e7-9f29-70cd8de6be9a.png)

**Comment by minggo** [comment]
It is because the background color is black, so the effect is a little different. If you set the background to white, then you will see the effect. For example, modify the codes like
```c++
    Director::getInstance()->setClearColor(Color4F(255,255,255,1)); // add this line the set background color to white
    
    auto director = Director::getInstance();
    auto origin = director->getVisibleOrigin();
    auto size = director->getVisibleSize();
    Vec2 center(origin.x + size.width/2, origin.y + size.height/2);
    _layer = LayerRadialGradient::create(Color4B(145, 106, 209, 140), Color4B(0,0,0,1), 200.0f, center, 0);
    addChild(_layer);
```

**Comment by Anywhere-Music-Player** [comment]
But I'm rendering only radial layer, and save png file to compare here, so it's whatever what bg color is. And I compare them both here on github by posting links and on white bg by the way :)

**Comment by minggo** [comment]
alpha:0.55 = 255 * 0.55 = 140, not 180.

**Comment by minggo** [comment]
`render->saveToFile("123.png", Image::Format::PNG);`, the result is `radial gradient` blend with `background color`. So make sure the background color is (0,0,0,0), can use `director->setClearColor(0,0,0,0)` to make sure.

**Comment by Anywhere-Music-Player** [comment]
I wrote why is that before

**Comment by minggo** [comment]
@KAMIKAZEUA you can use `LayerGradientColor::setBlend()` to try different blend mode, such as 

```c++
_layer->setBlendFunc(BlendFunc(GL_ONE, GL_ONE_MINUS_SRC_ALPHA));
```
More blend function can refer to [this doc](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glBlendFunc.xhtml).

LayerGradientColor uses `{GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA}` by default.

**Comment by minggo** [comment]
@KAMIKAZEUA what's the effect of your Objc version?

**Comment by Anywhere-Music-Player** [comment]
This https://github.com/cocos2d/cocos2d-x/issues/17480#issuecomment-302318405

**Comment by minggo** [comment]
It seems objc version doesn't interpolation the alpha value. You can change the shader(`ccShader_LayerRadialGradient.frag`)
from

```
gl_FragColor = mix(u_startColor, u_endColor, (d - u_expand) / (1.0 - u_expand));
```
to

```
gl_FragColor = vec4(mix(u_startColor, u_endColor, (d - u_expand) / (1.0 - u_expand)).rgb, u_startColor.a);
```

**Comment by Anywhere-Music-Player** [comment]
@minggo I've changed ccShader_LayerRadialGradient.frag as you suggested and used this code: 

```
Director::getInstance()->setClearColor(Color4F(255,255,255,1));
_layer = LayerRadialGradient::create(Color4B(145, 106, 209, 200), Color4B(0,0,0,0), 350.0f, center, 0);
addChild(_layer);
RenderTexture *render = RenderTexture::create(size.width, size.height);
render->begin();
_layer->visit();
render->end();
render->saveToFile("123.png", Image::Format::PNG);
```
Result is almost what I need: 

![1](https://cloud.githubusercontent.com/assets/3391456/26198604/5200201c-3bcf-11e7-862c-1adad7aa4466.png)

Except it's using black color.. and not transparent now?

Setting doesn't change anything:     
```
BlendFunc blendFunc = { GL_ONE, GL_ONE_MINUS_SRC_ALPHA };
_layer->setBlendFunc(blendFunc);
```

**Comment by minggo** [comment]
@KAMIKAZEUA how did you get the png file? I can not get full png file on Mac.

I think you can do it like:

* revert the shader
* pass `Color4B(0,0,0,200)` to end color: the opacity is the same as begin

Edit: i mean, the png file on Mac just part of the circle, not a full circle.

**Comment by Anywhere-Music-Player** [comment]
@minggo it's just a print screen.

**Comment by minggo** [comment]
I see, i use the codes you paste above. The result is error on Mac since render texture has bug on Mac with retina display.

**Comment by Anywhere-Music-Player** [comment]
How to revert a shader? 
Passing 200 to end color gives me this(also I'm playing a bit with opacity now it's 180 here): 
![123](https://cloud.githubusercontent.com/assets/3391456/26205618/fd7aaec4-3bea-11e7-8534-34175df66511.png)

End color and opacity should be 0, thats for sure.
Anyways, this gradient looks almost good as needed, except it goes to black.
p.s. I'm using Ctrl+Shift+4 on my hackintosh to create a screen-shoot.

**Comment by Anywhere-Music-Player** [comment]
It's like should be something mix of this code :) 
`gl_FragColor = mix(u_startColor, u_endColor, (d - u_expand) / (1.0 - u_expand));`
and this 
`gl_FragColor = vec4(mix(u_startColor, u_endColor, (d - u_expand) / (1.0 - u_expand)).rgb, u_startColor.a);` 

Because first code gives opacity goes to 0 and second is that what I need, but it goes to wrong color..

**Comment by minggo** [comment]
The difference is whether to interpolate opacity. The opacity will be used when doing blending. Change blend function to `{GL_ONE, GL_ONE_MINUS_SRC_ALPHA }` should have effect. What platform did you test on?

**Comment by Anywhere-Music-Player** [comment]
```
Director::getInstance()->setClearColor(Color4F(255,255,255,1));
_layer = LayerRadialGradient::create(Color4B(145, 106, 209, 180), Color4B(0,0,0,0), 300.0f, center, 0);
addChild(_layer);
BlendFunc blendFunc = { GL_ONE, GL_ONE_MINUS_SRC_ALPHA };
_layer->setBlendFunc(blendFunc);
```

![screen shot 2017-05-18 at 5 21 21 pm](https://cloud.githubusercontent.com/assets/3391456/26206819/78a29474-3bee-11e7-8b4c-10cf00a9be43.png)

**Comment by Anywhere-Music-Player** [comment]
Just tested on iPhone. 
Only this code: 
```
    _layer = LayerRadialGradient::create(Color4B(145, 106, 209, 180), Color4B(0,0,0,0), 300.0f, center, 0);
    addChild(_layer);
```
Output: 
![1](https://cloud.githubusercontent.com/assets/3391456/26207476/5154ff36-3bf0-11e7-9f4b-270d7762fef3.png)

This code: 

```
    Director::getInstance()->setClearColor(Color4F(255,255,255,1));
    _layer = LayerRadialGradient::create(Color4B(145, 106, 209, 180), Color4B(0,0,0,0), 300.0f, center, 0);
    addChild(_layer);
```
Output: 

![2](https://cloud.githubusercontent.com/assets/3391456/26207515/67dd7968-3bf0-11e7-8179-fc5f2edf1d1d.png)

And this:
```
    Director::getInstance()->setClearColor(Color4F(255,255,255,1));
    _layer = LayerRadialGradient::create(Color4B(145, 106, 209, 180), Color4B(0,0,0,0), 300.0f, center, 0);
    addChild(_layer);
    
    BlendFunc blendFunc = { GL_ONE, GL_ONE_MINUS_SRC_ALPHA };
    _layer->setBlendFunc(blendFunc);
```

Output: 

![3](https://cloud.githubusercontent.com/assets/3391456/26207574/976f3c52-3bf0-11e7-92d1-cf8f3dbef3e9.png)

As reminder, this was tested with: 

`gl_FragColor = vec4(mix(u_startColor, u_endColor, (d - u_expand) / (1.0 - u_expand)).rgb, u_startColor.a);`

**Comment by minggo** [comment]
Oh, i think you end color should be Color4B(255, 255, 255, 0);

**Comment by Anywhere-Music-Player** [comment]
Ehhm.. are you sure? I need transparent color at the end, but not white. 
In my game behind this gradient can be many objects with many other colors. So endColor can be only Color4B(0, 0, 0, 0).

**Comment by Anywhere-Music-Player** [comment]
Also, same code works differently on macOS and iOS. 
Running code:
```
    Director::getInstance()->setClearColor(Color4F(255,255,255,1));
    _layer = LayerRadialGradient::create(Color4B(145, 106, 209, 180),Color4B(0, 0, 0, 0), 300.0f, center, 0.0f);

```

![ios](https://cloud.githubusercontent.com/assets/3391456/26214300/8eadbb52-3c04-11e7-8d41-46488e633896.png)

![screen shot 2017-05-18 at 8 06 49 pm](https://cloud.githubusercontent.com/assets/3391456/26214572/8ae548f4-3c05-11e7-917c-e6c741a418bd.png)

**Comment by Anywhere-Music-Player** [comment]
Also, I'm trying to implement this to https://github.com/newnon/cocos2d-x which I'm using with https://github.com/newnon/SpriteBuilderX.

Question 1: `GLViewImpl::getFrameSize()`  for newnon/cocos2d-x it's cause problems with coordinates on macOS.. don't know why, but when I removed this override - and all works as before. Tested removed override on your cocos2d-x, Is this really needed?

Question 2: can this gradient be coordinated(located) in own local layer space in pixels? But not using screen size or visible size as currently? So, I can use it with Node and place it at any position and maybe resize with parent?

**Comment by minggo** [comment]
> Ehhm.. are you sure? I need transparent color at the end, but not white.
In my game behind this gradient can be many objects with many other colors. So endColor can be only Color4B(0, 0, 0, 0).

The opacity determine it transparent or not, not the color component, and (0, 0, 0) is black. There are two components will be interpolated: the color component(r,g,b) and opacity(a).

> Also, same code works differently on macOS and iOS.

I found the radius calculation has some issue, i will fix it.

> Question 1: GLViewImpl::getFrameSize() for newnon/cocos2d-x it's cause problems with coordinates on macOS.. don't know why, but when I removed this override - and all works as before. Tested removed override on your cocos2d-x, Is this really needed?

I need to get windows size in pixels to calculate the center and radius. Because in fragment shader, all are measured in pixel. Yep, it may break compatibility, i will find other way to do it.

> Question 2: can this gradient be coordinated(located) in own local layer space in pixels? But not using screen size or visible size as currently? So, I can use it with Node and place it at any position and maybe resize with parent?

cocos2d-x not use pixels, it use coordinate based on design resolution. As you see, when you set `center` to the center of design resolution, it is in the center of design resolution(480,320) in cpp-tests.

**Comment by minggo** [comment]
@KAMIKAZEUA you are right. I used local coordinate now. You can try the test case again, you can change the attribute in the test case, such as change the scale, skewx attribute that belongs to Node, and it will take effect.

**Comment by Anywhere-Music-Player** [comment]
@minggo Everything works as expected so far 👍 . Tested on iOS and macOS. Thank you very much. 
I hope on android all will be ok too, but I can't test this currently. Maybe only on emulator.

**Comment by minggo** [comment]
@KAMIKAZEUA good to hear that, it should work on Android too.

**Comment by Anywhere-Music-Player** [comment]
@minggo I got some problem today... don't know why I haven't get this before... 
So, just creating gradient with this code in tests:

```
Vec2 center(origin.x + size.width, origin.y + size.height/2);
_layer = LayerRadialGradient::create(Color4B(145, 106, 209, 180),Color4B(0, 0, 0, 0), 300.0f, center, 1.0f);
```
gives me this: 

![photo 6-7-17 23 54 00](https://user-images.githubusercontent.com/3391456/26900989-cff8b8b8-4bdc-11e7-88d5-9d7533c9ccc8.png)


But with radius 255, output is: 

![photo 6-7-17 23 52 51](https://user-images.githubusercontent.com/3391456/26900999-d60e8192-4bdc-11e7-8987-eb6859189244.png)


If radius more than 255 it's just fills whole screen. Very strange, but it's only happens on real device. I've tested on iPhone 6. But on Mac & Simulator everything is as it should be..
